### DISCUSSION TITLE
meaning of function flush()

### DISCUSSION THREAD

**Discussion by AlbertoPresta** [InterDigitalInc/CompressAI#205] [Q&A]
Hi, 

What is the meaning of function    encoder.flush() in line 644 of google.py module? 

Thanks in advance for answer, 
Alberto

**Comment by YodaEmbedding** [comment]
[Flushing](https://stackoverflow.com/a/15042890/365102) a buffer means emptying its contents, *and using these contents to do some work*. For instance, try running:

```python
import subprocess

with open("file.txt", "w") as f:
    print("This line was flushed.", file=f)
    f.flush()

    print("$ cat file.txt")
    subprocess.run(["cat", "file.txt"])

    print("This line will not show until flushed.", file=f)
    print("This line is also buffered.", file=f)

    print("$ cat file.txt")
    subprocess.run(["cat", "file.txt"])

    # f.flush()  # Automatically called when file is closed, or if buffer grows too large.
```

Output:
```bash
$ cat file.txt
This line was flushed.
$ cat file.txt
This line was flushed.
```

In CompressAI, the underlying encoder is buffering the intermediate symbols that it will encode. When `encoder.flush()` is called, it converts all these into an output bitstream. In google.py, we're not really using its buffering capabilities much since we immediately flush it after a single call to `encode_with_indexes` (which corresponds to "print" in the example above).
